File: Elixir/hello_world_again/lib/ecc_library/ecc_library.ex
defmodule EccLibrary do
  def start_link() do
    EccLibraryServer.start_link(__MODULE__, [])
  end

  def init(_) do
    []
  end

  def borrow_book(pid, name) do
    EccLibraryServer.call(pid, {:borrow, name})
  end

  def return_book(pid, name) do
    EccLibraryServer.cast(pid, {:return, name})
  end

  def close_library(pid), do: EccLibraryServer.call(pid, :terminate)

  def handle_call({:borrow, name}, from, library) do
    if Enum.empty?(library) do
      EccLibraryServer.reply(from, "ECC " <> name)
      library
    else
      EccLibraryServer.reply(from, hd(library))
      tl(library)
    end
  end

  def handle_call(:terminate, from, library) do
    EccLibraryServer.reply(from, :ok)
    terminate(library)
  end

  def handle_cast({:return, name}, library) do
    [name | library]
  end

  def terminate(library) do
    for s <- library, do: IO.puts("#{s} was burnt.")
    exit(:normal)
  end
end

File: Erlang/hello_world/src/ecc_library/ecc_library.erl
-module(ecc_library).
-export([start_link/0, borrow_book/2, return_book/2, close_library/1]).
-export([handle_cast/2, handle_call/3, init/1]).

start_link() -> ecc_library_server:start_link(?MODULE, []).

init([]) ->
    [].

borrow_book(Pid, Name) ->
    ecc_library_server:call(Pid, {borrow, Name}).

return_book(Pid, Name) ->
    ecc_library_server:cast(Pid, {return, Name}).

close_library(Pid) ->
    ecc_library_server:call(Pid, terminate).

handle_call({borrow, Name}, From, Library) ->
    if Library =:= [] ->
        ecc_library_server:reply(From, unicode:characters_to_list(["ECC ", Name], unicode)),
        Library;
       Library =/= [] ->
        ecc_library_server:reply(From, hd(Library)),
        tl(Library)
    end;

handle_call(terminate, From, Library) ->
    ecc_library_server:reply(From, ok),
    terminate(Library).
        
handle_cast({return, Book}, Library) ->
    [Book | Library].

terminate(Library) ->
    [io:format("~p was burnt.~n", [Name]) || Name <- Library],
    exit(normal).